src/librustc/ty/layout.rs

```diff
@@ -546,7 +546,8 @@ pub struct Struct {
 
     pub min_size: Size,
 
-    /// Size of padding needed by the struct due to requested repr alignment
+    /// Size of padding needed by the struct due to requested repr alignment.
+    /// This is created as required and will be empty if the structs fields require no padding.
     pub padding: Vec<Size>,
```

Comment:
You mean replacing the Struct offset vec with (field_offset, field_size) tuple and calculating padding on demand when the LLVM type is created? This on it's own wouldn't remove the need for `memory_index_with_padding` though and translating the index to include padding fields would get more expensive. 

I did think about always emitting padding even for primitive aligned types but I wasn't sure what impact that might have on LLVM internals, if that means a lot more work for it to do. Is it possible to emit a zero sized field? Generally I've taken a cautious approach and only changed behaviour if repr align is being used.

BTW, the `memory_index_with_padding_fields` is now only `O(n)` if repr align is being used.